### PR TITLE
editor: count chars and words in textarea

### PR DESCRIPTION
**Pull request by sebastiendeleze**
* Adds counters in textarea for chars and words, depending on the configuration.
* Avoids to exceed the counts when typing.
* Checks for new resource context to avoid `pid` to be set in model.

Co-Authored-by: Sébastien Délèze <sebastien.deleze@rero.ch>

### CODE REVIEW

File: projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts [hunk 1]
```diff
@@ -0,0 +1,138 @@
...
+  template: `
+    <textarea
+      [formControl]="formControl"
+      [cols]="to.cols"
+      [rows]="to.rows"
+      class="form-control"
+      [class.is-invalid]="showError"
+      [formlyAttributes]="field"
+    >
+    >
```
Comment (Garfield-fr) [inline projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts:36]: Remove `>`

File: projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts [hunk 2]
```diff
@@ -0,0 +1,138 @@
...
+    </textarea
+    >
```
Comment (Garfield-fr) [inline projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts:38]: `>` on the same line as textarea

File: projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts [hunk 3]
```diff
@@ -0,0 +1,138 @@
...
+        field.templateOptions.limitChars || field.templateOptions.displayChars
+      "
+    >
+      <span class="small text-muted d-inline-block"
+        >{{ 'Number of chars' | translate }}: {{ countChars
+        }}<ng-container *ngIf="field.templateOptions.limitChars">
+          / {{ field.templateOptions.limitChars }}</ng-container
+        ></span
+      >
+    </ng-container>
```
Comment (Garfield-fr) [inline projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts:62]: Why you don't use `ng-template` with a parameter `display` only with values `word` or `char`

File: projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts [hunk 4]
```diff
@@ -0,0 +1,138 @@
...
+  `,
+  changeDetection: ChangeDetectionStrategy.OnPush,
+})
+export class TextareaFieldComponent
+  extends FormlyFieldTextArea
+  implements OnInit {
+  /**
```
Comment (Garfield-fr) [inline projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts:69]: space before

File: projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts [hunk 5]
```diff
@@ -0,0 +1,138 @@
...
+    return this.formControl.value.split(' ').length;
+  }
+
+  /**
+   * Component init.
+   *
+   * Adds validator for chars and words.
+   * Check if the current value does not exceed the limit value for chars or words.
+   */
+  ngOnInit() {
```
Comment (Garfield-fr) [inline projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts:101]: `ngOnInit(): void {`

**Review by Garfield-fr — CHANGES_REQUESTED** [review]
Reformat the html part `template`

File: projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts [hunk 6]
```diff
@@ -0,0 +1,138 @@
...
+      [formlyAttributes]="field"
+    >
+    >
+    </textarea
+    >
+    <ng-container
+      *ngIf="
+        field.templateOptions.limitWords || field.templateOptions.displayWords
+      "
+    >
```
Comment (zannkukai) [inline projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts:43]: Why not one a single line ? (as it used ini all other component (when possible))

File: projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts [hunk 7]
```diff
@@ -0,0 +1,138 @@
...
+      <span class="small text-muted d-inline-block mr-3"
+        >{{ 'Number of words' | translate }}: {{ countWords }}
+        <ng-container *ngIf="field.templateOptions.limitWords">
+          / {{ field.templateOptions.limitWords }}</ng-container
+        ></span
+      >
```
Comment (zannkukai) [inline projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts:49]: Little bit confused to read XD

File: projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts [hunk 8]
```diff
@@ -0,0 +1,138 @@
...
+   * Get the number of chars.
+   *
+   * @returns The number of chars.
+   */
+  get countChars(): number {
+    if (!this.formControl.value) {
+      return 0;
+    }
+
+    return this.formControl.value.length;
```
Comment (zannkukai) [inline projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts:79]: ternary ? 
```
return (!this.formControl.value)
  ? 0
  : this.formControl.value.length;
```

File: projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts [hunk 9]
```diff
@@ -0,0 +1,138 @@
...
+   * Get the number of words.
+   *
+   * @returns The number of words.
+   */
+  get countWords(): number {
+    if (!this.formControl.value) {
+      return 0;
+    }
+
+    return this.formControl.value.split(' ').length;
```
Comment (zannkukai) [inline projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts:92]: ternary ?

**Comment by sebastiendeleze** [comment]
@zannkukai, @Garfield-fr I use `prettier` for formatting with `printWidth` configuration to `80` (default value). And that's the result... Which formatter are you using?

File: projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts (hunk 3 above)
Comment (sebastiendeleze) [reply projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts:62]: I'm not sure to understand, could you be more precise, please?

**Comment by zannkukai** [comment]
> @zannkukai, @Garfield-fr I use `prettier` for formatting with `printWidth` configuration to `80` (default value). And that's the result... Which formatter are you using?

We don't use a 80 characters line limit in angular projects. Isn't it ?
For me this code isn't readable for an human
```
      <span class="small text-muted d-inline-block"
        >{{ 'Number of chars' | translate }}: {{ countChars
        }}<ng-container *ngIf="field.templateOptions.limitChars">
          / {{ field.templateOptions.limitChars }}</ng-container
        ></span
      >
 ```
 It seems (always for me) better to write
 ```
 <span class="small text-muted d-inline-block">
     {{ 'Number of chars' | translate }}: {{ countChars }}
     <ng-container *ngIf="field.templateOptions.limitChars">
          / {{ field.templateOptions.limitChars }}
     </ng-container>
</span>
```
 
 At least have the `>` on the same line than the tag

File: projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts (hunk 3 above)
Comment (sebastiendeleze) [reply projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts]: OK, I changed the way to handle this.

File: projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts [hunk 10]
```diff
@@ -0,0 +1,109 @@
...
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+import { ChangeDetectionStrategy, Component, OnInit } from '@angular/core';
+import { AbstractControl, ValidatorFn } from '@angular/forms';
+import { FormlyFieldTextArea } from '@ngx-formly/bootstrap';
+
+@Component({
+  selector: 'ng-core-editor-formly-field-textarea',
+  template: `
+    <textarea
```
Comment (jma) [inline projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts:24]: Can be in a separate file.

File: projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts [hunk 11]
```diff
@@ -0,0 +1,109 @@
...
+      class="form-control"
+      [class.is-invalid]="showError"
+      [formlyAttributes]="field"
+    ></textarea>
+    <ng-container
+      [ngTemplateOutlet]="counter"
+      [ngTemplateOutletContext]="{
+        limit: field.templateOptions.limitWords,
+        count: countWords,
+        label: 'Number of words'
```
Comment (jma) [inline projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts:37]: Translations?

File: projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts [hunk 12]
```diff
@@ -0,0 +1,109 @@
...
+      }"
+      *ngIf="field.templateOptions.limitWords || field.templateOptions.displayWords"
+    ></ng-container>
+    <ng-container
+      [ngTemplateOutlet]="counter"
+      [ngTemplateOutletContext]="{
+        limit: field.templateOptions.limitChars,
+        count: countChars,
+        label: 'Number of chars'
```
Comment (jma) [inline projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts:46]: Translations?

File: projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+      *ngIf="field.templateOptions.limitChars || field.templateOptions.displayChars"
+    ></ng-container>
+    <ng-template #counter let-limit="limit" let-count="count" let-label="label">
+      <span class="small text-muted d-inline-block mr-3">
+        {{ label | translate }}: {{ count }}
+        <ng-container *ngIf="limit"> / {{ limit }} </ng-container>
+      </span>
+    </ng-template>
+  `,
+  changeDetection: ChangeDetectionStrategy.OnPush,
```
Comment (jma) [inline projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts:57]: Perhaps a small comment to explain why?

File: projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts [hunk 14]
```diff
@@ -0,0 +1,109 @@
...
+    return !this.formControl.value ? 0 : this.formControl.value.length;
+  }
+
+  /**
+   * Get the number of words.
+   *
+   * @returns The number of words.
+   */
+  get countWords(): number {
+    return !this.formControl.value ? 0 : this.formControl.value.split(' ').length;
```
Comment (jma) [inline projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts:75]: I would prefers something like: `split(/\s+/)`

**Review by jma — APPROVED** [review]

File: projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts (hunk 14 above)
Comment (sebastiendeleze) [reply projects/rero/ng-core/src/lib/record/editor/type/textarea/textarea.component.ts]: Absolutely!

**Pull request merged**